python/pyspark/sql/dataframe.py

```diff
@@ -30,6 +30,7 @@
     overload,
     TYPE_CHECKING,
 )
+from typing_extensions import Concatenate, ParamSpec
```

Comment:
Isn't this a thirdparty library? We should probably wait till we drop old Python versions

Reply:
Thanks for taking the time to review!

I've moved this import to its corresponding `_typing.pyi` file, as is done elsewhere in Spark when using objects that come from `typing_extensions`:

https://github.com/apache/spark/blob/97e0a88b1d04a8e33206e5cb8384878ca4cdfc5a/python/pyspark/mllib/stat/_statistics.py#L30-L31

https://github.com/apache/spark/blob/97e0a88b1d04a8e33206e5cb8384878ca4cdfc5a/python/pyspark/mllib/_typing.pyi#L31

https://github.com/apache/spark/blob/97e0a88b1d04a8e33206e5cb8384878ca4cdfc5a/python/pyspark/mllib/_typing.pyi#L21